core/src/main/java/io/grpc/inprocess/InProcessServer.java

```diff
@@ -71,15 +81,17 @@ static InProcessServer findServer(String name) {
   public void start(ServerListener serverListener) throws IOException {
     this.listener = serverListener;
     this.scheduler = schedulerPool.getObject();
-    // Must be last, as channels can start connecting after this point.
-    if (registry.putIfAbsent(name, this) != null) {
-      throw new IOException("name already registered: " + name);
+    if (!anonymous) {
+      // Must be last, as channels can start connecting after this point.
+      if (registry.putIfAbsent(name, this) != null) {
+        throw new IOException("name already registered: " + name);
+      }
     }
   }
 
   @Override
   public SocketAddress getListenSocketAddress() {
-    return new InProcessSocketAddress(name);
+    return new InProcessSocketAddress(name, anonymous ? this : null);
```

Comment:
This is creating a new InProcessSocketAddress each method call. We don't have to make the address 1:1, but I lean that direction. In any case, either this (canonical address) or InProcessSocketAddress (to use server's hashcode/equals) would need to change.